py2exe/runtime.py

```diff
@@ -417,10 +417,7 @@ def build_archive(self, libpath, delete_existing_resources=False):
         # keys; we only need one of them in the archive.
         for mod in set(self.mf.modules.values()):
             if mod.__code__:
-                if hasattr(mod, "__path__"):
-                    path = mod.__name__.replace(".", "\\") + "\\__init__" + bytecode_suffix
-                else:
-                    path = mod.__name__.replace(".", "\\") + bytecode_suffix
+                path =mod.__compiled_file__

```

Comment:
This will not fall back to the old behavior if e.g. `source` is not `None` or `self.__file__` exists. Perhaps that's the desired behavior though.

Reply:
It should not fallback to this behavior when __source__ is None, that's correct, since there is in fact no source that we can recompile with the new file name in place. There is actually nothing that really changes here, only the logic to calculate the compiled path is moved from the runtime to mf3.Module.

Could you elaborate about "if __file__ exists? There is a check for __file__, but that's only when there is no source, so that's also perfectly valid here I believe.

Reply:
Is just that the old code is not executed in case `hasattr(self, __file__)`is True. This discrepancy made me wonder if it could cause problems or not.

Reply:
The old code should still always be executed if there is no source. Just the order changed from precompiled > source to source > precompiled.
For example, we could make __compiled_file__ return None or always raise an error when there is no source, and then make the runtime print the exception.
If you have any suggestions to make the current code look clearer, please let me know.

Reply:
I am talking about this branch:
```
if source is None:
    if hasattr(self, __file__):
         return self.__file__
```
that prevents this to run:
```
        if hasattr(self, "__path__"):
	            return self.__name__.replace(".", "\\") + "\\__init__" + bytecode_suffix
	        else:
	            return self.__name__.replace(".", "\\") + bytecode_suffix
```
so that `path =mod.__compiled_file__` in this case will be different than in the old code even if source is None.

Reply:
Ah yes, I do now understand your point. I will change this.